### ISSUE TITLE
How to deal with the need for Nx4 data of some materials

### ISSUE THREAD

**Issue by almarklein** [pygfx/pygfx#42]
## Intro

With vertex buffers, if a shader input is vec4, and the vbo has Nx2, the z and w element will be zero. This works, because for vertex buffers we provide additional information about the striding of the data. With storage buffers (aka SSBO) we just have some bytes that read from/write to in the shader. This means the data must match with what the shader expects.

There is also a thing with vec3's being padded to 16 bytes. Things are confusing:
*  It looks to me like the [Khronos docs](https://www.khronos.org/opengl/wiki/Interface_Block_(GLSL)#Memory_layout) say that this padding does not occur when using std430 memory layout.
* But if I create a glsl shader with a vec3-array buffer and compile it to SpirV, then I see it's ArrayStride always being 16.
* I added some code in PyShader to annotate the array's stride to 12, but that does not help (the context gets lost in both cases).
* also see [this](https://www.khronos.org/opengl/wiki/Interface_Block_(GLSL)#Memory_layout)

I think that even if we do find a way to use vec3-arrays in storage buffers, we probably should not, because I would not feel confident that all drivers support it. Plus it might cause trouble for people who want to write shaders in glsl.

## The situation

In short, the rendering of some materials (e.g. lines) require the buffers data to match with the shader, and cannot use vec3. 

## Options

More a list of ideas. We may use a combination of these:

* Test and require these restrictions in the render functions where they're needed.
* Require these restrictions always.
* Disallow Nx3 arrays.
* Render functions to which this applies can also implement both 2D (vec2) and 3D (vec4) versions.
* Autoconvert data. E.g. positions and colors always become Nx4 when doing `Geometry(positions=[[1,2], [3,4], [5,6]])`
* Add padding for Nx3 data when it's uploaded.
* Provide a function that makes it explicit but easy to convert data.
* The 4th element can be used for other means.

**Comment by almarklein** [comment]
Mmm, instead of using `(pyshader.RES_BUFFER, (0, 2), Array(vec4))`, we can use an array of `f32`, and sample each element from the 1D array in the shader. If we go that route, we could e.g. force all position data to be 3D (which would be alright, I think).

**Comment by almarklein** [comment]
We basically did what I suggested in the last comment.
